server/blob/blob-union/src/main/java/org/apache/james/blob/union/HybridBlobStore.java

```diff
@@ -128,45 +145,53 @@ private HybridBlobStore(BlobStore lowCostBlobStore, BlobStore highPerformanceBlo
 
     @Override
     public Mono<BlobId> save(BucketName bucketName, byte[] data, StoragePolicy storagePolicy) {
-        return selectBlobStore(storagePolicy, Mono.just(data.length > configuration.getSizeThreshold()))
-            .flatMap(blobStore -> blobStore.save(bucketName, data, storagePolicy));
+        return save(bucketName, new ByteArrayInputStream(data), storagePolicy);
     }
 
     @Override
     public Mono<BlobId> save(BucketName bucketName, InputStream data, StoragePolicy storagePolicy) {
         Preconditions.checkNotNull(data);
 
-        BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
-        return selectBlobStore(storagePolicy, Mono.fromCallable(() -> isItABigStream(bufferedInputStream)))
-            .flatMap(blobStore -> blobStore.save(bucketName, bufferedInputStream, storagePolicy));
-    }
-
-    private Mono<BlobStore> selectBlobStore(StoragePolicy storagePolicy, Mono<Boolean> largeData) {
         switch (storagePolicy) {
             case LOW_COST:
-                return Mono.just(lowCostBlobStore);
+                return lowCostBlobStore.save(bucketName, data, storagePolicy);
+
             case SIZE_BASED:
-                return largeData.map(isLarge -> {
-                    if (isLarge) {
-                        return lowCostBlobStore;
-                    }
-                    return highPerformanceBlobStore;
-                });
-            case HIGH_PERFORMANCE:
-                return Mono.just(highPerformanceBlobStore);
+                BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
+                return Mono.fromCallable(() -> isItABigStream(bufferedInputStream))
+                    .filter(Boolean::booleanValue)
+                    .flatMap(ignored -> lowCostBlobStore.save(bucketName, bufferedInputStream, storagePolicy))
+                    .switchIfEmpty(Mono.defer(() ->
+                        checkingDuplicateWriteThenSave(bucketName, readAllBytes(bufferedInputStream), storagePolicy)));
```

Comment:
is Mono.defer() necessary here? The checkingDuplicateWriteThenSave() method already returns a Mono.

Reply:
Maybe not. Can we try to remove it and see if it breaks?

Reply:
A gree